### PR TITLE
Always encode arrays as integer-indexed hashes

### LINKED ISSUE #561: GET requests with arrays in data not handled correctly by stripe API
I noticed (and found a workaround to solve) a bug occurring in GET requests with data containing arrays.

For example: `stripe.invoices.retrieve('upcoming', {..., subscription_items: [..., ...] })`

This will try to make a GET request with url-encoded data (in the body of the request apparently(?)), but the converted querystring doesn't* have array indexes.
* purposely, according to the commit found via blame ; and also according to the 'Stripe-specific cases'  test in utils.spec.js)

Thing is: Stripe API seems to need them to understand the query.
Current code in utils.js:
```javascript
  /**
   * Stringifies an Object, accommodating nested objects
   * (forming the conventional key 'parent[child]=value')
   */
  stringifyRequestData: function(data) {
    // console.log('stringifyRequestData', data, qs.stringify(data, {arrayFormat: 'brackets'}))
    return qs.stringify(data, {arrayFormat: 'brackets'});
  },
```
outputs:
```
subscription=sub_ES6MM9DZExqcN0&subscription_prorate=true&subscription_items%5B%5D%5Bid%5D=si_ES6MwymN93s0Fh&subscription_items%5B%5D%5Bdeleted%5D=true&subscription_items%5B%5D%5Bplan%5D=business_monthly&subscription_items%5B%5D%5Bquantity%5D=1&coupon=redacted
```
urldecoded:
```
subscription=sub_ES6MM9DZExqcN0&subscription_prorate=true&subscription_items[][id]=si_ES6MwymN93s0Fh&subscription_items[][deleted]=true&subscription_items[][plan]=business_monthly&subscription_items[][quantity]=1&coupon=redacted
```
Log trace in Stripe Dashboard: 
screenshot at 02-05-39

workaround:
```javascript
  /**
   * Stringifies an Object, accommodating nested objects
   * (forming the conventional key 'parent[child]=value')
   */
  stringifyRequestData: function(data) {
    // console.log('stringifyRequestData', data, qs.stringify(data, {arrayFormat: 'brackets'}))
    return qs.stringify(data);
  },
```
outputs:
```
subscription=sub_ES6MM9DZExqcN0&subscription_prorate=true&subscription_items%5B0%5D%5Bid%5D=si_ES6MwymN93s0Fh&subscription_items%5B0%5D%5Bdeleted%5D=true&subscription_items%5B1%5D%5Bplan%5D=business_monthly&subscription_items%5B1%5D%5Bquantity%5D=1&coupon=redacted
```
urldecoded:
```
subscription=sub_ES6MM9DZExqcN0&subscription_prorate=true&subscription_items[0][id]=si_ES6MwymN93s0Fh&subscription_items[0][deleted]=true&subscription_items[1][plan]=business_monthly&subscription_items[1][quantity]=1&coupon=redacted
```
Log trace in Stripe Dashboard: 
screenshot at 02-06-32

So, either you have to fix it in your API backend (to accept unindexed array backets), or fix it here (to provide API compatible format).

Thanks.

### PR DESCRIPTION
**Pull request by ob-stripe**
r? @brandur-stripe @rattrayalex-stripe 
cc @stripe/api-libraries 

Always encode arrays as integer-indexed hashes. This is the default behavior of `qs.stringify()`. Stripe's API can now accept integer-indexed hashes for all array parameters (including `expand`) so there is no longer a need to special-case arrays of hashes parameters like `subscription_items`.

I've also updated the encoding method to change square brackets back to their literals to make the strings easier to read. This is consistent with most of our other libraries.

Fixes #561.

### CODE REVIEW

File: test/resources/Invoices.spec.js [hunk 1]
```diff
@@ -137,18 +137,11 @@ describe('Invoices Resource', function() {
...
-              2: {
-                deleted: true,
-                id: 'SOME_ID',
-              },
-            },
+            subscription_items: [
+              {plan: 'potato'},
+              {plan: 'rutabaga'},
+              {id: 'SOME_ID', deleted: true},
+            ],
```
Comment (rattrayalex-stripe) [inline test/resources/Invoices.spec.js]: I'm somewhat worried about our test coverage reduction here... is there anything in our test suite to verify that these POST bodies are form-encoded with `subscription_items[0][plan]` instead of `subscription_items[][plan]` ?

File: test/utils.spec.js [hunk 2]
```diff
@@ -34,33 +34,33 @@ describe('utils', function() {
...
 
     it('Handles arrays of objects', function() {
-      expect(decodeURI(utils.stringifyRequestData({
+      expect(utils.stringifyRequestData({
         a: [
           {b: 'c'},
           {b: 'd'},
         ],
-      }))).to.equal('a[][b]=c&a[][b]=d');
+      })).to.equal('a[0][b]=c&a[1][b]=d');
```
Comment (rattrayalex-stripe) [inline test/utils.spec.js]: \o/

File: test/utils.spec.js [hunk 3]
```diff
@@ -34,33 +34,33 @@ describe('utils', function() {
...
-      expect(decodeURI(utils.stringifyRequestData({
+      expect(utils.stringifyRequestData({
         a: {
-          0: 'c',
-          1: 'd',
+          0: {b: 'c'},
+          1: {b: 'd'},
         },
-      }))).to.equal('a[0]=c&a[1]=d');
+      })).to.equal('a[0][b]=c&a[1][b]=d');
```
Comment (rattrayalex-stripe) [inline test/utils.spec.js]: curious, why this change?

File: test/utils.spec.js [hunk 4]
```diff
@@ -76,18 +76,10 @@ describe('utils', function() {
...
       ].join('&'));
     });
-
-    describe('Stripe-specific cases', function() {
-      it('Handles the `expand` array correctly (producing the form `expand[]=_` for each item', function() {
-        expect(decodeURI(utils.stringifyRequestData({
-          expand: ['a', 'foo', 'a.b.c'],
-        }))).to.equal('expand[]=a&expand[]=foo&expand[]=a.b.c');
-      });
-    });
```
Comment (rattrayalex-stripe) [inline test/utils.spec.js]: why remove this test?

Comment (rattrayalex-stripe) [reply test/utils.spec.js]: on that note... we probably _do_ ideally want want `expand` to continue to be `[]` instead of `[0]`, right? Not a big deal, but it is prettier 😄

File: test/utils.spec.js (hunk 3 above)
Comment (ob-stripe) [reply test/utils.spec.js]: Mainly to illustrate that a list of hashes and an integer-indexed hash of hashes have the same encoded representation.

File: test/utils.spec.js (hunk 4 above)
Comment (ob-stripe) [reply test/utils.spec.js]: My preference would be to avoid unnecessary special-casing and use the same encoding scheme everywhere. Which is why I removed this test, there's no real need to test "Stripe-specific" stuff since there are no special cases anymore.

File: test/resources/Invoices.spec.js (hunk 1 above)
Comment (ob-stripe) [reply test/resources/Invoices.spec.js]: Fair! Looking into it.

Comment (ob-stripe) [reply test/resources/Invoices.spec.js]: Okay, so this is difficult to test in the current state. The URL-encoding of the request parameters happens in the original `_request` method, but `getSpyableStripe()` returns an instance of `stripe` where `_request` is patched to memorize the method parameters instead of actually sending the request.

We've recently added `nock` for testing the telemetry stuff, so we could probably write some new tests that use it to check the actual body of our HTTP requests.

File: test/utils.spec.js (hunk 4 above)
Comment (rattrayalex-stripe) [reply test/utils.spec.js]: Fair enough - I assume we have this behavior in our other libs now too, so best to be consistent. 👍

**Comment by rattrayalex-stripe** [comment]
LGTM

File: test/resources/Invoices.spec.js (hunk 1 above)
Comment (rattrayalex-stripe) [reply test/resources/Invoices.spec.js]: Yeah, I think nock sounds like the right tool for this job, should be pretty straightforward - fft DM me or paul if you need help with it!

Comment (ob-stripe) [reply test/resources/Invoices.spec.js]: So I lied, we didn't add nock at all. No idea why I thought that was the case!

Nevertheless, I tried writing a test with nock to test `StripeResource._request`, but I can't seem to have nock intercept the request -- for some reason the request is always sent to the real API server. Mind giving it a try @rattrayalex-stripe?

Comment (rattrayalex-stripe) [reply test/resources/Invoices.spec.js]: We add nock in https://github.com/stripe/stripe-node/pull/559 😄

**Comment by ob-stripe** [comment]
@rattrayalex-stripe I added a couple of encoding tests in `StripeResource.spec.js`, ptal.

Unfortunately, for GET requests, it seems nock doesn't let you check the raw query string -- instead, it uses `qs` itself to decode the query string into an object. I'd have preferred to check the encoded string directly just in case `qs`' logic differs from Stripe's in subtle ways, but that doesn't seem to be possible.

File: test/StripeResource.spec.js [hunk 5]
```diff
@@ -36,6 +36,64 @@ describe('StripeResource', function() {
...
+          host: stripe.getConstant('DEFAULT_HOST'),
+          path: '/v1/subscriptions/sub_123',
+          data: {
+            customer: 'cus_123',
+            items: [
+              {plan: 'foo', quantity: 2},
+              {id: 'si_123', deleted: true},
+            ],
+          },
+          body: 'customer=cus_123&items[0][plan]=foo&items[0][quantity]=2&items[1][id]=si_123&items[1][deleted]=true',
```
Comment (rattrayalex-stripe) [inline test/StripeResource.spec.js]: just be crystal-clear because I don't see an `expect` and I'm not super-familiar with `nock`, but if we wrote `[]` instead of `[1]` here, the test fails, right?

Comment (ob-stripe) [reply test/StripeResource.spec.js]: Right! The call to `scope.done()` asserts that all specified calls on that scope were performed.

**Comment by ob-stripe** [comment]
Released as 6.25.1.
